File: src/main/java/core/procedures/io/Println.java
package core.procedures.io;

import core.Repl;
import core.exceptions.SCMIOException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.OutputPort;
import core.scm.Void;
import core.writer.Writer;

import java.io.IOException;

public final class Println extends AFn {

  public Println() {
    super(new FnArgsBuilder().min(1).max(2).mandatory(new Class[]{Object.class})
                             .rest(OutputPort.class).build());
  }

  @Override
  public String getName() {
    return "println";
  }

  @Override
  public Object apply(Object... args) {
    OutputPort outputPort;
    if (args.length == 1) {
      outputPort = Repl.getCurrentOutputPort();
    } else {
      outputPort = ((OutputPort)args[1]);
    }
    Object arg = args[0];
    try {
      if ((arg instanceof CharSequence) || (arg instanceof Character)) {
        outputPort.writeln(arg.toString());
      } else {
        outputPort.writeln(Writer.write(arg));
      }
    } catch (IOException e) {
      throw new SCMIOException(e);
    }
    return Void.VOID;
  }
}

File: src/main/java/core/procedures/io/Println.kt
package core.procedures.io

import core.Repl
import core.procedures.AFn
import core.scm.OutputPort
import core.Writer
import core.procedures.Arity.Range

class Println : AFn<Any?, Unit>(name = "println", arity = Range(1, 2),
                                mandatoryArgsTypes = arrayOf(Any::class.java), restArgsType = OutputPort::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        1    -> Repl.currentOutputPort
        else -> args[1] as OutputPort
    }.let {
        when (args[0]) {
            is CharSequence, is Char -> it.writeln(args[0].toString())
            else -> it.writeln(Writer.write(args[0]))
        }
    }
}
